runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/MapControlClientPool.java

```diff
@@ -0,0 +1,99 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.fnexecution.control;
+
+import com.google.common.collect.Maps;
+import java.time.Duration;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * A {@link ControlClientPool} backed by a client map. It is expected that a given client id will be
+ * requested at most once.
+ */
+public class MapControlClientPool implements ControlClientPool {
+
+  /** Creates a {@link MapControlClientPool} with an unspecified timeout. */
+  public static MapControlClientPool create() {
+    return withTimeout(Duration.ofSeconds(Long.MAX_VALUE));
+  }
+
+  /**
+   * Creates a {@link MapControlClientPool} with the given timeout. Timeouts only apply to source
+   * requests.
+   */
+  public static MapControlClientPool withTimeout(Duration timeout) {
+    return new MapControlClientPool(timeout);
+  }
+
+  private final Duration timeout;
+  private final Object lock = new Object();
+
+  @GuardedBy("lock")
+  private final Map<String, CompletableFuture<InstructionRequestHandler>> clients =
+      Maps.newHashMap();
+
+  private MapControlClientPool(Duration timeout) {
+    this.timeout = timeout;
+  }
+
+  @Override
+  public ControlClientSource getSource() {
+    return this::getClient;
+  }
+
+  @Override
+  public ControlClientSink getSink() {
+    return this::putClient;
+  }
+
+  private void putClient(String workerId, InstructionRequestHandler client) {
+    synchronized (lock) {
+      CompletableFuture<InstructionRequestHandler> future =
+          clients.computeIfAbsent(workerId, MapControlClientPool::createClientFuture);
+      boolean success = future.complete(client);
+      if (!success) {
+        throw new IllegalStateException(
+            String.format("Control client for worker id %s failed to compete", workerId));
+      }
+    }
+  }
+
+  private InstructionRequestHandler getClient(String workerId)
+      throws ExecutionException, InterruptedException, TimeoutException {
+    CompletableFuture<InstructionRequestHandler> future;
+    synchronized (lock) {
+      future = clients.computeIfAbsent(workerId, MapControlClientPool::createClientFuture);
+    }
+    // TODO: Wire in health checking of clients so requests don't hang.
+    future.get(timeout.getSeconds(), TimeUnit.SECONDS);
```

Comment:
<!--thread_id:cc_182896323_t; commit:1fd121da1417624b3b84f0300648251da64b9cb5; resolved:1-->
<!--section:context-quote-->
> **jkff** wrote:
> What exactly are we waiting for here: are we waiting for a matching put(workerId, _) call? When can there happen a getClient(workerId) call for a workerId that hasn't been put() yet?

<!--section:body-->
`put()` is generally called when the control server receives an incoming connection. This happens out-of-band from the `get()` and is actually _usually_ expected to occur after the `get()` due to container startup latency.